src/inmanta/deploy/scheduler.py

```diff
@@ -0,0 +1,163 @@
+"""
+
+import asyncio
+from collections.abc import Mapping, Set
+from typing import Optional
+
+from inmanta.data.model import ResourceIdStr
+from inmanta.deploy import work
+from inmanta.deploy.state import ModelState, ResourceDetails, ResourceStatus
+
+
+# TODO: name
+# TODO: expand docstring
+class Scheduler:
+    """
+    Scheduler for resource actions. Reads resource state from the database and accepts deploy, dry-run, ... requests from the
+    server. Schedules these requests as tasks according to priorities and, in case of deploy tasks, requires-provides edges.
+    """
+    def __init__(self) -> None:
+        self._state: ModelState = ModelState(version=0)
+        self._work: work.ScheduledWork = work.ScheduledWork(requires=model_state.requires)
+
+        # We uphold two locks to prevent concurrency conflicts between external events (e.g. new version or deploy request)
+        # and the task executor background tasks.
+        #
+        # - lock to block scheduler state access (both model state and scheduled work) during scheduler-wide state updates
+        #   (e.g. trigger deploy). A single lock suffices since all state accesses (both read and write) by the task runners are
+        #   short, synchronous operations (and therefore we wouldn't gain anything by allowing multiple readers).
+        # TODO: this lock may have little value left now. Consider.
+        self._scheduler_lock: asyncio.Lock = asyncio.Lock()
+        # - lock to serialize scheduler state updates (i.e. process new version)
+        self._update_lock: asyncio.Lock = asyncio.Lock()
+
+    def start(self) -> None:
+        # TODO (ticket): read from DB instead
+        pass
+
+    async def deploy(self) -> None:
+        async with self._scheduler_lock:
+            # TODO: more efficient access to dirty set by caching it on the ModelState
+            dirty: Set[ResourceIdStr] = {
+                r for r, details in self._state.resource_state.items() if details.status == ResourceStatus.HAS_UPDATE
+            }
+            # TODO: pass in running deploys
+            self._work.update_state(ensure_scheduled=dirty, running_deploys={})
+
+    # TODO: name
+    # TODO (ticket): design step 2: read new state from DB instead of accepting as parameter (method should be notification only, i.e. 0 parameters)
+    async def new_version(
+        self,
+        version: int,
+        resources: Mapping[ResourceIdStr, ResourceDetails],
+        requires: Mapping[ResourceIdStr, Set[ResourceIdStr]],
+    ) -> None:
+        async with self._update_lock:
+            # Inspect new state and mark resources as "update pending" where appropriate. Since this method is the only writer
+            # for "update pending", and a stale read is acceptable, we can do this part before acquiring the exclusive scheduler
+            # lock.
+            # TODO: what to do when an export changes handler code without changing attributes? Consider in deployed state? What
```

Comment:
desires state is considered to be defined by the attributes and nothing else. 

Given that updates are rare, we can keep the behavior like that. 
(Apart from that argument, I would keep the behavior as it is, but I'm not going to linger on this)